### ISSUE TITLE
about distriuted training

### ISSUE THREAD

**Issue by Xjmengnieer** [noahzn/Lite-Mono#43]
Hello author, I have made modifications according to your prompts to achieve distributed training!

but But the following very strange phenomenon occurred：

This is my startup command
![image](https://github.com/noahzn/Lite-Mono/assets/88227513/dc564b25-b2f2-4702-ac65-e6eeebf2b291)

And these are the modified code
![image](https://github.com/noahzn/Lite-Mono/assets/88227513/e83e91f0-b4c7-4405-93ee-ab548a00fb03)
![image](https://github.com/noahzn/Lite-Mono/assets/88227513/2ebc11bd-5d9e-4854-bafa-939a293ec71c)

The following is the usage of the GPUS
![image](https://github.com/noahzn/Lite-Mono/assets/88227513/c620d334-3fd9-4490-8083-4662d7dd16e1)

It can be clearly seen that card 0 is occupying too much space！！

I don't know what went wrong. I hope you can help me

**Comment by Xjmengnieer** [comment]
And I found through debugging that the problem occurred here in DDP, and when each card was subjected to DDP, the number 0 card would significantly increase:

this is before DDP:
![image](https://github.com/noahzn/Lite-Mono/assets/88227513/3e81873f-7029-4deb-a1a8-513f7dba746c)

this is 0 and 1 gpus after DDP:
![image](https://github.com/noahzn/Lite-Mono/assets/88227513/cfc4c0c2-7f0e-4283-848e-044b66b73d2c)

and this is  all 0,1,2 after DDP:

![image](https://github.com/noahzn/Lite-Mono/assets/88227513/76df74c0-d25a-4d6f-b502-51019623a347)

it seems like the model have been copied to 0 gpu!!

**Comment by noahzn** [comment]
Hi, please check if all the losses are computed on GPU0.

**Comment by Xjmengnieer** [comment]
> Hi, please check if all the losses are computed on GPU0.

no, I debug step by step，They are on their respective gpus：

The following is a scenario where only three gpus are used.

As can be seen, the occupancy of card 0 is still much higher than that of other cards, but it is proportional to the number of cards
![image](https://github.com/noahzn/Lite-Mono/assets/88227513/b45923b4-a5dd-4089-af0e-a157308ca1a4)

**Comment by Xjmengnieer** [comment]
The occupancy of gpu 0 has been higher than other gpus since DDP execution, And it seems that everything is normal during the training process

![image](https://github.com/noahzn/Lite-Mono/assets/88227513/4751e29c-a10e-473a-9d40-e9d512470610)

**Comment by noahzn** [comment]
Yes. There is a lot of discussion on the internet about unbalanced GPU usage with DDP.

**Comment by Xjmengnieer** [comment]
> Yes. There is a lot of discussion on the internet about unbalanced GPU usage with DDP.

Can it be because your model is built in blocks, rather than a complete model?

**Comment by noahzn** [comment]
I don't think so. The full name of DDP is Distributed Data Parallel, so you are actually parallelizing your data, not the model.

**Comment by Xjmengnieer** [comment]
> I don't think so. The full name of DDP is Distributed Data Parallel, so you are actually parallelizing your data, not the model.

but now, The problem seems to have nothing to do with the data，

**Comment by noahzn** [comment]
https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel
`This container provides data parallelism by synchronizing gradients across each model replica. `
`To use DistributedDataParallel on a host with N GPUs, you should spawn up N processes, ensuring that each process exclusively works on a single GPU from 0 to N-1.`

**Comment by Xjmengnieer** [comment]
Yes, I have complied with this regulation：

![image](https://github.com/noahzn/Lite-Mono/assets/88227513/6244bb7b-94e5-4f89-a289-6bce19c79170)

**Comment by Xjmengnieer** [comment]
If possible, could you also try implementing a distributed training script to see if the same problem will occur.

be deeply grateful

**Comment by noahzn** [comment]
> If possible, could you also try implementing a distributed training script to see if the same problem will occur.
> 
> be deeply grateful

Unfortunately, I am currently working on other projects and do not have time to implement this feature. Maybe I will have a look at it when I have some free time. If you solve this problem, feel free to share your solutions.

**Comment by Xjmengnieer** [comment]
ok，thanks

**Comment by Xjmengnieer** [comment]
> 
I have identified such an error, but I am not sure which parameter of this stage is. Do you know
![image](https://github.com/noahzn/Lite-Mono/assets/88227513/e34effd9-2242-430c-8b83-9676f760cc2a)

**Comment by Xjmengnieer** [comment]
It should be in module depth_decoder

**Comment by Xjmengnieer** [comment]
Can't layer-norm perform distributed training

**Comment by noahzn** [comment]
Hi, is it the LayerNorm used in the `LGFI` module? Have you modified other parts of the structure?

I found [another issue](https://discuss.pytorch.org/t/parameters-which-did-not-receive-grad-for-rank/144150) that might be related.

**Comment by Xjmengnieer** [comment]
> Hi, is it the LayerNorm used in the `LGFI` module? Have you modified other parts of the structure?
> 
> I found [another issue](https://discuss.pytorch.org/t/parameters-which-did-not-receive-grad-for-rank/144150) that might be related.

no, I havn't modified other parts of the structure ！ I think operation layer_norm itself does not involve distributed communication，Therefore, when conducting DDP distributed training, “find_unused_parameters” must be set to True

**Comment by noahzn** [comment]
Yes, it's possible. See this [link](https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html#torch.nn.LayerNorm).

`This layer uses statistics computed from input data in both training and evaluation modes.` I don't know if this is the problem.

**Comment by Xjmengnieer** [comment]
Operation layer-norm is indeed a problem. After I commented it out, a new problem appeared, and it seems that there was an omission in the calculation of the loss function

![image](https://github.com/noahzn/Lite-Mono/assets/88227513/d754baa2-011c-4e11-9b5e-37364e37cfd6)

**Comment by noahzn** [comment]
Are you loading the ImageNet pre-trained weights? The `fc` layer seems to be the last unused layer in the pre-trained model.

**Comment by Xjmengnieer** [comment]
It should be，and the occupancy of card 0 is still much higher than that of other cards, and we still don't know where the reason lies

**Comment by owl-of-pastos** [comment]
Hi, I'm also trying to achieve distributed training and my code is just like the issue's author's. I recently encountered a warning that I'd like to inquire about. The warning message is as follows:

> /home/dingyl/miniconda3/envs/litemono/lib/python3.7/site-packages/torch/autograd/__init__.py:175: UserWarning: Grad strides do not match bucket view strides. This may indicate grad was not created according to the gradient layout contract, or that the param's strides changed since DDP was constructed.  This is not an error, but may impair performance.
> grad.sizes() = [48, 64, 1, 1], strides() = [64, 1, 64, 64]
> bucket_view.sizes() = [48, 64, 1, 1], strides() = [64, 1, 1, 1] (Triggered internally at  /opt/conda/conda-bld/pytorch_1659484809535/work/torch/csrc/distributed/c10d/reducer.cpp:312.)
>   allow_unreachable=True, accumulate_grad=True)  # Calls into the C++ engine to run the backward pass

Could you kindly provide some insights into what might be causing this warning？
Furthermore, I haven't encountered high GPU0 occupancy issues, but DDP seems to have a minimal impact on training speed.

**Comment by noahzn** [comment]
> UserWarning: Grad strides do not match bucket view strides. This may indicate grad was not created according to the gradient layout contract, or that the param's strides changed since DDP was constructed.

Hi @owl-of-pastos, [this](https://github.com/pytorch/pytorch/issues/47163) might be helpful.
